app/controllers/users_controller.rb

```diff
@@ -17,7 +17,7 @@ def new
   def create
     @user = User.new user_params
     if user.save
-      user.send_activation_email
+      UserMailer.account_activation(user).deliver_now
```

Comment:
cơ chế hoạt động của Mailer là gì?
chạy từ đâu đến đâu?

Reply:
1. generate mailer
Ví dụ generate User mailer
 rails generate mailer UserMailer 
2. Edit mailer: thêm các action và sử dụng view để cấu trúc nội dung trong đó controller thì tạo ra các file html để gửi về client thì mailer tạo ra các message để gửi về email
Ví dụ: 
  def account_activation user
    @user = user
    mail to: user.email, subject: t("account_activation")
  end
Method này nhận 2 giá trị là to-để chỉ định địa chỉ sẽ được gửi đến và subject-nội dung của subject thư
3. Tạo mailer view 
Tạo file với cấu trúc tên là account_activation.html.erb trong thư mục view/user_mailer
 File này chính là template sẽ được sử dụng cho email.
Cũng có thể tạo email dạng text với tên account_activation.text.erb
4. Gọi mailer 
  UserMailer.account_activation(user).deliver_now
Phương thức account_activation trả về đối tượng ActionMailer::MessageDelivery mà sau đó có thể được cho biết deliver_now(gửi ngay) hoặc deliver_later(email không đựoc gửi ngay khi hàm được gọi) để send nó đi. Đối tượng ActionMailer::MessageDelivery chỉ là một gói bao quanh một Mail::Message

Reply:
chị hỏi cách hoạt động nữa mà? không phải step by step để tạo nó nhé em yêu

Reply:
UserMailer.account_activation(user).deliver_now
UserMailer chính là class UserMailer ạ 
trong class UserMailer thì chứa action ở mailer đó. cụ thể ở đây là account_activation(user) được sử dụng trong câu lệnh trên
mà class UserMailer lại được kế thừa từ class ApplicationMailer. ApplicationMailer thì được kế thừa từ ActionMailer::Base, cho phép gửi email từ ứng dụng của bằng cách sử dụng mailer model và view
Khi sử dụng UserMailer.account_activation(user) thì trả về ActionMailer::MessageDelivery object để tạo mail. 
Lớp ActionMailer::MessageDelivery là một gói bao quanh Mail::Message.